### ISSUE TITLE
Proposal: c8d: expose contentstore API

### ISSUE THREAD

**Issue by tonistiigi** [moby/moby#44369]
### Description

Currently, the only way to load and save images in/from Docker is to use `docker save/load` commands that require a tarball to be created that contains all layers and manifests of the image.

For example, when doing builds and loading the build result into Docker, this full tarball needs to be created by the builder and transferred to Docker which then extracts it on disk before processing it. When doing these builds repeatedly, usually only a couple of these layers actually change but because of the inefficiency of this API full tarball needs to be transferred every time even though eventually Docker will discard most of it.

A similar case appears when you would want to transfer an image from one Docker instance to another(without a registry in between). If the target Docker instance already has the older version of the image and just needs an updated version it is again likely that only one layer actually needs to be exported from the daemon and imported on the other side. But no API exists today that would allow running this efficiently.

With the move to containerd image backend that is currently in progress, we have some new options to solve this issue. I propose exposing containerd's contentstore API https://pkg.go.dev/github.com/containerd/containerd@v1.6.9/api/services/content/v1#ContentServer so it can be used with Docker, but do it in a safe manner.

These APIs allow creating individual blobs in content-addressable storage, checking if blob with a specific digest already exists and pulling the blob's contents.

We can use the existing `/grpc` endpoint that can be upgraded to gRPC connection(currently used by BuildKit).

Containerd APIs on their own are unsafe, and calling them with incorrect parameters can completely corrupt the internal storage, delete objects that are in use by other users, or leak objects forever, so they never release their storage. These cases should be avoided by the Docker API.

Before client would call the content API methods it would need to make a call that creates a **lease** Leases and GC labels are how containerd controls what objects can be deleted by the GC. All the subsequent content API calls need to contain a lease ID and will error otherwise. A lease has a timeout (for example 10 min by default). If client needs more time to finish its transactions, it can ask for the timeout to be extended. If the client crashes and terminates the connection to the daemon all objects associated with the lease get their reference count reduced when the timeout is reached. If the reference count reaches 0, the object is deleted by containerd GC.

With the lease, client can check which blobs daemon already knows about. If it needs to, it can upload the newer blobs that remain associated with the lease. Blobs containing the image manifest and config are uploaded in the same way. Finally, client will make a `CreateImage()` call that contains the digest of the root manifest(either image manifest or manifest index for multi-platform). This method will validate that the digest points to a valid manifest with all layers present in contentstore and create a new record under `docker images`. Now all the blobs are associated with the image itself and the lease is not needed anymore to keep the blobs from getting deleted.

For safety, the `Delete()` method in contentstore API in Moby is completely disabled. Only way to manage storage is to create leases and delete them.

These APIs enable clients to efficiently create new images when the daemon already knows about some of the layers(iterative builds, deployments) as well as only export the layers the client needs out of the image instead of downloading the full `docker save` tarball.

@rumpl @AkihiroSuda

**Comment by AkihiroSuda** [comment]
SGTM

**Comment by tianon** [comment]
I really like this!  Do you imagine this will require creating Moby/Docker-specific wrappers for the containerd API methods, or would we be able to use the standard containerd client implementation directly against the "upgraded" socket connection?  (ie, for an existing containerd client implementation, how much "change" is required to talk to Docker's content store vs talking to containerd directly? probably not much beyond the initial connection code/negotiation, right?)

This idea sounds like it's fleshed out well enough that you might already have a PoC somewhere? :eyes:

**Comment by tonistiigi** [comment]
@tianon We can use the containerd's API definitions directly. On the dockerd side, there would be a wrapper, similar to https://github.com/moby/buildkit/blob/master/session/content/attachable.go#L122-L123 that would do the extra validation and if then redirect the call to https://pkg.go.dev/github.com/containerd/containerd#Client.ContentStore .

**Comment by thaJeztah** [comment]
Slightly related;

- https://github.com/moby/moby/issues/8039
- https://github.com/moby/moby/issues/19718

**Comment by tianon** [comment]
I've said it privately several times, so I should probably post it publicly: I think about this proposal on a very regular basis now -- having ready access to a containerd content/image store would be really, really interesting and useful, and we discussed in today's maintainer meeting how having something like this would allow users to use `ctr` directly to query/inspect the underlying guts of the containerd content store (in the context of #45122 and a discussion around "inspecting pointers" vs "inspecting pointees"), which would mean we don't necessarily need to write as much of that in a higher-level CLI/API for users to be able to access/query/browse/understand this data (or at least, not as _urgently_).

**Comment by jabrown85** [comment]
Heavy plus one on this. 

I work on the [buildpacks](https://github.com/buildpacks) project and today we use [a tar trick](https://github.com/buildpacks/imgutil/blob/main/local/save.go#L112-L119) to try and prevent extracting and then re-sending layers that exist in the daemon storage. This method greatly reduces the time spent re-building a slightly modified image over and over. 

However, this trick crashes docker desktop on the mac when you swap to containerd back with an EOF error. We should probably put in an issue for this (?).

For our project in particular we are looking at a rather large increase in build times for our users when the switch to containerd storage with no obvious workaround. If we could access the storage described through here perhaps we could end up in a situation where builds are actually faster as we would be able to more accurately skip work on subsequent rebuilds and not transfer in and out of the daemon via `docker load|save`.

**Comment by edmorley** [comment]
We desperately need this API - currently without it we do not have a way to use the containerd image store in a performant way - which causes a significant performance regression when using it over the classic store:
https://github.com/buildpacks/pack/issues/2272

**Comment by BwL1289** [comment]
> We desperately need this API 

Seconded. Is there any estimated timeline?

**Comment by BwL1289** [comment]
Bumping this. In addition to all of the other use-cases (i.e. sending tarball stuck), the dev experience and added CI/CD build times makes this worthwhile.

**Comment by rafikk** [comment]
Is this actually completed? https://github.com/moby/moby/pull/50744#discussion_r2610252548 states that the containerd content store is exposed as read-only which wouldn't satisfy the use cases in this thread.

**Comment by akerouanton** [comment]
No, it was closed automatically when I merged #50744. I'll reopen it.

**Comment by BwL1289** [comment]
Bumping this. Been 4 years.

**Comment by thaJeztah** [comment]
It somewhat works already (read-only); 

- see https://github.com/moby/moby/issues/51698
- and https://github.com/moby/moby/pull/50744#discussion_r2610252548

```bash
docker pull -q hello-world

ctr -a /var/run/docker.sock -n moby content ls
WARN[0000] Failed to check deprecations                  error="unknown service containerd.services.introspection.v1.Introspection"
DIGEST									SIZE	AGE		LABELS
sha256:5099b89d7666cc2186cad769ddc262ddc7c335b33f5fe79f9ffe50a01282b23e	1.027kB	6 seconds	containerd.io/distribution.source.docker.io=library/hello-world,containerd.io/gc.ref.content.config=sha256:eb84fdc6f2a3a064445bb2a2fbc89c515666c428d6c96b6ab68a4cd218819688,containerd.io/gc.ref.content.l.0=sha256:58dee6a49ef1c01bb8a00180d70f55b3527c8e7326a05b3c5135c4ff60cfb6d6,containerd.io/gc.ref.content.referrer.sha256.670045644a15=sha256:670045644a15cf4c09e24a61b924660c58f45c4c03c2d17d49b774c04c9bb62b
sha256:58dee6a49ef1c01bb8a00180d70f55b3527c8e7326a05b3c5135c4ff60cfb6d6	3.149kB	5 seconds	containerd.io/distribution.source.docker.io=library/hello-world,containerd.
...
```

**Comment by BwL1289** [comment]
> It somewhat works already (read-only);
> 
> * see [Explicitly expose content store gRPC service #51698](https://github.com/moby/moby/issues/51698)
> * and [Natively support gRPC on the docker socket #50744 (comment)](https://github.com/moby/moby/pull/50744#discussion_r2610252548)
> 
> docker pull -q hello-world
> 
> ctr -a /var/run/docker.sock -n moby content ls
> WARN[0000] Failed to check deprecations                  error="unknown service containerd.services.introspection.v1.Introspection"
> DIGEST									SIZE	AGE		LABELS
> sha256:5099b89d7666cc2186cad769ddc262ddc7c335b33f5fe79f9ffe50a01282b23e	1.027kB	6 seconds	containerd.io/distribution.source.docker.io=library/hello-world,containerd.io/gc.ref.content.config=sha256:eb84fdc6f2a3a064445bb2a2fbc89c515666c428d6c96b6ab68a4cd218819688,containerd.io/gc.ref.content.l.0=sha256:58dee6a49ef1c01bb8a00180d70f55b3527c8e7326a05b3c5135c4ff60cfb6d6,containerd.io/gc.ref.content.referrer.sha256.670045644a15=sha256:670045644a15cf4c09e24a61b924660c58f45c4c03c2d17d49b774c04c9bb62b
> sha256:58dee6a49ef1c01bb8a00180d70f55b3527c8e7326a05b3c5135c4ff60cfb6d6	3.149kB	5 seconds	containerd.io/distribution.source.docker.io=library/hello-world,containerd.
> ...

That does not satisfy most of the use cases on this thread. See https://github.com/moby/moby/issues/44369#issuecomment-3645970468
